## One Hot Encoding

In [3]:
# import necessary libraries
import pandas as pd
import numpy as np

In [5]:
# load dataset
data = pd.read_csv('../datasets/mercedesbenz_train.csv')
data.head()

,ID,y,X0,X1,X2,X3,X4,X5,X6,X8,...,X375,X376,X377,X378,X379,X380,X382,X383,X384,X385
0,0,130.81,k,v,at,a,d,u,j,o,...,0,0,1,0,0,0,0,0,0,0
1,6,88.53,k,t,av,e,d,y,l,o,...,1,0,0,0,0,0,0,0,0,0
2,7,76.26,az,w,n,c,d,x,j,x,...,0,0,0,0,0,0,1,0,0,0
3,9,80.62,az,t,n,f,d,x,l,e,...,0,0,0,0,0,0,0,0,0,0
4,13,78.02,az,v,n,f,d,h,d,n,...,0,0,0,0,0,0,0,0,0,0


In [7]:
data = data[['X1', 'X2', 'X3', 'X4', 'X5', 'X6']]
data.head()

,X1,X2,X3,X4,X5,X6
0,v,at,a,d,u,j
1,t,av,e,d,y,l
2,w,n,c,d,x,j
3,t,n,f,d,x,l
4,v,n,f,d,h,d


In [9]:
# let's have a look at how many labels each variable has
for col in data.columns:
    print(col, ': ', len(data[col].unique()), ' labels')

X1 :  27  labels
X2 :  44  labels
X3 :  7  labels
X4 :  4  labels
X5 :  29  labels
X6 :  12  labels


In [11]:
# let's examine how many columns we will obtain after one hot encoding these variables
pd.get_dummies(data, drop_first=True).shape

(4209, 117)

We can see that from just 6 initial categorical variables, we end up with 117 new variables.

What we can do instead?

In the winning solution of the KDD 2009 cup: "Winning the KDD Cup Orange Challenge with Ensemble Selection the authors limit one hot encoding to the 10 most frequent labels of the variable. This means that they would make one binary variable for each of the 10 most frequent labels only. This is equivalent to group all the other labels under a new category, that in this case will be dropped. Thus, the 10 new dummy variables indicate if one of the 10 most frequent labels is present (1) or not (0) for a particular observation."

In [13]:
# let's find top 10 most frequent categories for the variable X2
data.X2.value_counts(ascending=False).head(20)

X2
as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
ag      19
z       19
Name: count, dtype: int64

In [15]:
# let's make a list with the most frequent categories of the variable
top_10 = [x for x in data.X2.value_counts(ascending=False).head(10).index]
top_10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [17]:
# and now we make the 10 binary variables
for label in top_10:
    data[label] = np.where(data['X2'] == label, 1, 0)

data[['X2']+top_10].head(40)

,X2,as,ae,ai,m,ak,r,n,s,f,e
0,at,0,0,0,0,0,0,0,0,0,0
1,av,0,0,0,0,0,0,0,0,0,0
2,n,0,0,0,0,0,0,1,0,0,0
3,n,0,0,0,0,0,0,1,0,0,0
4,n,0,0,0,0,0,0,1,0,0,0
5,e,0,0,0,0,0,0,0,0,0,1
6,e,0,0,0,0,0,0,0,0,0,1
7,as,1,0,0,0,0,0,0,0,0,0
8,as,1,0,0,0,0,0,0,0,0,0
9,aq,0,0,0,0,0,0,0,0,0,0


In [19]:
data.columns

Index(['X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'as', 'ae', 'ai', 'm', 'ak', 'r',
       'n', 's', 'f', 'e'],
      dtype='object')

In [25]:
# get whole set of dummy variables, for all the categorical variables

def one_hot_top_x(df, variable, top_x_labels):
    # function to create the dummy variables for the most frequent labels
    # we can vary the number of most frequent labels that we encode

    for label in top_x_labels:
        df[variable+'_'+label] = np.where(data[variable]==label, 1, 0)

# read the data again
data = pd.read_csv('../datasets/mercedesbenz_train.csv')
display(data.head())
data = data[['X1', 'X2', 'X3', 'X4', 'X5', 'X6']]
display(data.head())

,ID,y,X0,X1,X2,X3,X4,X5,X6,X8,...,X375,X376,X377,X378,X379,X380,X382,X383,X384,X385
0,0,130.81,k,v,at,a,d,u,j,o,...,0,0,1,0,0,0,0,0,0,0
1,6,88.53,k,t,av,e,d,y,l,o,...,1,0,0,0,0,0,0,0,0,0
2,7,76.26,az,w,n,c,d,x,j,x,...,0,0,0,0,0,0,1,0,0,0
3,9,80.62,az,t,n,f,d,x,l,e,...,0,0,0,0,0,0,0,0,0,0
4,13,78.02,az,v,n,f,d,h,d,n,...,0,0,0,0,0,0,0,0,0,0


,X1,X2,X3,X4,X5,X6
0,v,at,a,d,u,j
1,t,av,e,d,y,l
2,w,n,c,d,x,j
3,t,n,f,d,x,l
4,v,n,f,d,h,d


In [27]:
# encode X2 into the 10 most frequent categories
top_10_x1 = [x for x in data.X1.value_counts(ascending=False).head(10).index]
top_10_x2 = [x for x in data.X2.value_counts(ascending=False).head(10).index]
top_10_x3 = [x for x in data.X3.value_counts(ascending=False).head(10).index]
top_10_x4 = [x for x in data.X4.value_counts(ascending=False).head(10).index]
top_10_x5 = [x for x in data.X5.value_counts(ascending=False).head(10).index]
top_10_x6 = [x for x in data.X6.value_counts(ascending=False).head(10).index]

for var in ['X1', 'X2', 'X3', 'X4', 'X5', 'X6']:
    one_hot_top_x(data, var, f'top_10_{var}')
    data.drop(var, axis=1, inplace=True)
    print(data.columns)

Index(['X2', 'X3', 'X4', 'X5', 'X6', 'X1_t', 'X1_o', 'X1_p', 'X1__', 'X1_1',
       'X1_0', 'X1_X'],
      dtype='object')
Index(['X3', 'X4', 'X5', 'X6', 'X1_t', 'X1_o', 'X1_p', 'X1__', 'X1_1', 'X1_0',
       'X1_X', 'X2_t', 'X2_o', 'X2_p', 'X2__', 'X2_1', 'X2_0', 'X2_X', 'X2_2'],
      dtype='object')
Index(['X4', 'X5', 'X6', 'X1_t', 'X1_o', 'X1_p', 'X1__', 'X1_1', 'X1_0',
       'X1_X', 'X2_t', 'X2_o', 'X2_p', 'X2__', 'X2_1', 'X2_0', 'X2_X', 'X2_2',
       'X3_t', 'X3_o', 'X3_p', 'X3__', 'X3_1', 'X3_0', 'X3_X', 'X3_3'],
      dtype='object')
Index(['X5', 'X6', 'X1_t', 'X1_o', 'X1_p', 'X1__', 'X1_1', 'X1_0', 'X1_X',
       'X2_t', 'X2_o', 'X2_p', 'X2__', 'X2_1', 'X2_0', 'X2_X', 'X2_2', 'X3_t',
       'X3_o', 'X3_p', 'X3__', 'X3_1', 'X3_0', 'X3_X', 'X3_3', 'X4_t', 'X4_o',
       'X4_p', 'X4__', 'X4_1', 'X4_0', 'X4_X', 'X4_4'],
      dtype='object')
Index(['X6', 'X1_t', 'X1_o', 'X1_p', 'X1__', 'X1_1', 'X1_0', 'X1_X', 'X2_t',
       'X2_o', 'X2_p', 'X2__', 'X2_1', 'X2_0', 'X2_X', 'X2_2',

In [29]:
data.shape

(4209, 47)

## Advantages
- Starightforward to implement.
- Does not require hours of variable exploration.
- Does not expand massively the feature space (number of columns in the dataset).

## Disadvantages
- Does not add any information that may make the variable more predictive.
- Does not keep the information of the ignored labels.

Because it is not unusual that categorical variables have a few dominating categories and the remaining labels add mostly noise, this is a quite simple and straightforward approach that may be useful on many occassions.

It is worth noting that the top 10 variables is a totally arbitrary number. You could also choose the top 5 or top 20.

# Count or Frequency Encoding 
(What we did earlier section as well.)

This technique is for high cardinality.

Another way to refer to variables that have a multitude of categories, is to call them variables with high cardinality.

If we have categorical variables containing many multiple labels or high cardinality by using one hot encoding, we will expand the feature space dramatically.

One approach that is heavily used in Kaggle competitions, is to replace each label of the categorical variable by the count, this is the amount of times each label appears in the dataset. Or the frequency this is the percentage of observations within that category. the 2 are equivalent.

Let's see how this works:

In [35]:
df = pd.read_csv('../datasets/mercedesbenz_train.csv')
display(df.head())
df = df[['X1', 'X2']]
display(df.head())

,ID,y,X0,X1,X2,X3,X4,X5,X6,X8,...,X375,X376,X377,X378,X379,X380,X382,X383,X384,X385
0,0,130.81,k,v,at,a,d,u,j,o,...,0,0,1,0,0,0,0,0,0,0
1,6,88.53,k,t,av,e,d,y,l,o,...,1,0,0,0,0,0,0,0,0,0
2,7,76.26,az,w,n,c,d,x,j,x,...,0,0,0,0,0,0,1,0,0,0
3,9,80.62,az,t,n,f,d,x,l,e,...,0,0,0,0,0,0,0,0,0,0
4,13,78.02,az,v,n,f,d,h,d,n,...,0,0,0,0,0,0,0,0,0,0


,X1,X2
0,v,at
1,t,av
2,w,n
3,t,n
4,v,n


In [37]:
df.shape

(4209, 2)

In [39]:
# one hot encoding (without dropping first column)
pd.get_dummies(df).shape

(4209, 71)

In [41]:
len(df['X1'].unique())

27

In [43]:
len(df['X2'].unique())

44

In [49]:
for col in df.columns:
    print(col, ': ', len(df[col].unique()), ' labels')

X1 :  27  labels
X2 :  44  labels


In [51]:
# let's obtain count for each one of the labels in variable X2
# let's capture this in a dictionary that we can use to re-map the labels

df.X2.value_counts().to_dict()

{'as': 1659,
 'ae': 496,
 'ai': 415,
 'm': 367,
 'ak': 265,
 'r': 153,
 'n': 137,
 's': 94,
 'f': 87,
 'e': 81,
 'aq': 63,
 'ay': 54,
 'a': 47,
 't': 29,
 'k': 25,
 'i': 25,
 'b': 21,
 'ao': 20,
 'ag': 19,
 'z': 19,
 'd': 18,
 'ac': 13,
 'g': 12,
 'ap': 11,
 'y': 11,
 'x': 10,
 'aw': 8,
 'at': 6,
 'h': 6,
 'al': 5,
 'an': 5,
 'q': 5,
 'av': 4,
 'ah': 4,
 'p': 4,
 'au': 3,
 'am': 1,
 'j': 1,
 'af': 1,
 'l': 1,
 'aa': 1,
 'c': 1,
 'o': 1,
 'ar': 1}

In [53]:
# and now let's replace each label in X2 by its count

# first we make a dictionary that maps each label to the counts
df_frequency_map = df.X2.value_counts().to_dict()

In [55]:
# and now we replace X2 labels in the dataset df
df.X2 = df.X2.map(df_frequency_map)
df.head()

,X1,X2
0,v,6
1,t,4
2,w,137
3,t,137
4,v,137


## Advantage
- It is very simple to implement.
- Does not increase the feature dimensional space.

## Disadvantage
- If some of the labels have the same count, then they will be replaced with the same count and they will loose some valueable information.
- Adds somewhat arbitrary numbers, and therefore weights to the different labels, that may not be related to their predictive power.

# Ordinal Numbering Encoding or Label Encoding
Ordinal data is a categorical, statistical data type where the variables have natural, ordered categories and the distances between the categories is not known.

Categorical data is of two types: 1. Nominal, 2. Ordinal

For ex. : Nominal - Pen , Pencil, Eraser, Cow, Dog, Cat

Ordinal - Fantastic > Okay > Don't Like

          Good < Better < Best etc.

- Student's grade in an exam (A, B, C or Fail).
- Educational level, with the categories: Elementary school, High School, College Graduate, PHD ranked from 1 to 4.

When the categorical variable are ordinal, the most straightforward best appraoch is to replace the labels by some ordinal number based on the ranks.


In [64]:
import datetime

In [68]:
# create a variable with dates, and from that extract the weekday
# I create a list of dates with 20 days difference from today
# and then transform it into a dataframe

df_base = datetime.datetime.today()
df_date_list = [df_base - datetime.timedelta(days=x) for x in range(0, 20)]
df = pd.DataFrame(df_date_list)
df.columns = ['day']
df

,day
0,2024-07-20 22:31:05.551061
1,2024-07-19 22:31:05.551061
2,2024-07-18 22:31:05.551061
3,2024-07-17 22:31:05.551061
4,2024-07-16 22:31:05.551061
5,2024-07-15 22:31:05.551061
6,2024-07-14 22:31:05.551061
7,2024-07-13 22:31:05.551061
8,2024-07-12 22:31:05.551061
9,2024-07-11 22:31:05.551061


In [76]:
# extract the week day name
df['day_of_week'] = df['day'].dt.day_name()
df.head()

,day,day_of_week
0,2024-07-20 22:31:05.551061,Saturday
1,2024-07-19 22:31:05.551061,Friday
2,2024-07-18 22:31:05.551061,Thursday
3,2024-07-17 22:31:05.551061,Wednesday
4,2024-07-16 22:31:05.551061,Tuesday


In [78]:
# engineer categorical variable by ordinal number replacement

weekday_map = {'Monday' : 1,
              'Tuesday' : 2,
              'Wednesday' : 3,
              'Thursday' : 4,
              'Friday' : 5,
              'Saturday' : 6,
              'Sunday' : 7
              }
df['day_ordinal'] = df.day_of_week.map(weekday_map)
df.head()

,day,day_of_week,day_ordinal
0,2024-07-20 22:31:05.551061,Saturday,6
1,2024-07-19 22:31:05.551061,Friday,5
2,2024-07-18 22:31:05.551061,Thursday,4
3,2024-07-17 22:31:05.551061,Wednesday,3
4,2024-07-16 22:31:05.551061,Tuesday,2


In [ ]:
# Advantages of Orinal Measurement
# Ordinal Measurement is normally used for surveys and questionnaries. Statistical analysis is applied to the responses once they are collected to place the people who took the survey into the various categories. The data is then compared to draw inferences and conclusions about the whole surveyed population with regard to the specific variables.

In [88]:
a=40
b=5
if a>b:
    print("y")
else:
    print("n")

y
